### PR TITLE
Improvement in consecutives History::get_change [1.9.x]

### PR DESCRIPTION
**Pull request by richiware**
Added a mechanism to improve the performance with consecutive calls to `History::get_change()`.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1018/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/848/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1014/)
* Performance [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Performance%20Tests%20Localhost/317/)

File: src/cpp/rtps/reader/StatefulReader.cpp [hunk 1]
```diff
@@ -587,14 +587,20 @@ bool StatefulReader::processGapMsg(
...
         {
             if (pWP->irrelevant_change_set(auxSN))
             {
-                CacheChange_t* to_remove = findCacheInFragmentedProcess(auxSN, pWP->guid());
+                CacheChange_t* to_remove = nullptr;
+                auto ret_iterator = findCacheInFragmentedProcess(auxSN, pWP->guid(), &to_remove, history_iterator);
                 if (to_remove != nullptr)
                 {
-                    mp_history->remove_change(to_remove);
+                    ret_iterator = mp_history->remove_change_nts(to_remove, ret_iterator);
```
Comment (MiguelCompany) [inline src/cpp/rtps/reader/StatefulReader.cpp]: Perhaps it should be `history_iterator = mp_history->remove_change_nts(to_remove, ret_iterator);`

File: src/cpp/rtps/reader/StatefulReader.cpp [hunk 2]
```diff
@@ -603,10 +609,15 @@ bool StatefulReader::processGapMsg(
         {
             if (pWP->irrelevant_change_set(it))
             {
-                CacheChange_t* to_remove = findCacheInFragmentedProcess(auxSN, pWP->guid());
+                CacheChange_t* to_remove = nullptr;
+                auto ret_iterator = findCacheInFragmentedProcess(auxSN, pWP->guid(), &to_remove, history_iterator);
                 if (to_remove != nullptr)
                 {
-                    mp_history->remove_change(to_remove);
+                    ret_iterator = mp_history->remove_change_nts(to_remove, ret_iterator);
```
Comment (MiguelCompany) [inline src/cpp/rtps/reader/StatefulReader.cpp]: Perhaps it should be `history_iterator = mp_history->remove_change_nts(to_remove, ret_iterator);`

File: include/fastrtps/rtps/reader/RTPSReader.h [hunk 3]
```diff
@@ -233,11 +233,18 @@ class RTPSReader : public Endpoint
...
-     * @return If a CacheChange_t was found, it will be returned. In other case nullptr is returned.
+     * @param change If a CacheChange_t was found, this argument will fill with its pointer.
+     * In other case nullptr is returned.
+     * @param hint Iterator since the search will start.
+     * Used to improve the search.
+     * @return Iterator pointing to the position were CacheChange_t was found.
+     * It can be used to improve next search.
      */
-    CacheChange_t* findCacheInFragmentedProcess(
+    History::const_iterator findCacheInFragmentedProcess(
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastrtps/rtps/reader/RTPSReader.h]: Done

File: include/fastrtps/rtps/history/History.h [hunk 4]
```diff
@@ -189,6 +193,12 @@ class History
             const GUID_t& guid,
             CacheChange_t** change) const;
 
+    RTPS_DllAPI const_iterator get_change_nts(
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastrtps/rtps/history/History.h]: Done

File: src/cpp/rtps/reader/StatefulReader.cpp (hunk 1 above)
Comment (richiware) [reply src/cpp/rtps/reader/StatefulReader.cpp]: You're right, done

File: src/cpp/rtps/reader/StatefulReader.cpp (hunk 2 above)
Comment (richiware) [reply src/cpp/rtps/reader/StatefulReader.cpp]: You're right, done

**Pull request merged**